rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,189 @@
+        delayed = /* unspecifed */
+    };
+
+    task_arena(int max_concurrency = automatic, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               workers_leave a_workers_leave = workers_leave::delayed);
+
+    task_arena(const constraints& constraints_, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               workers_leave a_workers_leave = workers_leave::delayed);
+
+    void initialize(int max_concurrency, unsigned reserved_for_masters = 1,
+                    priority a_priority = priority::normal,
+                    workers_leave a_workers_leave = workers_leave::delayed);
+
+    void initialize(constraints a_constraints, unsigned reserved_for_masters = 1,
+                    priority a_priority = priority::normal,
+                    workers_leave a_workers_leave = workers_leave::delayed);
+
+    void start_parallel_block();
+    void end_parallel_block(bool set_one_time_fast_leave = false);
+
+    class scoped_parallel_block {
+        scoped_parallel_block(task_arena& ta, bool set_one_time_fast_leave = false);
+    };
+};
+
+namespace this_task_arena {
+    void start_parallel_block();
+    void end_parallel_block(bool set_one_time_fast_leave = false);
+}
+```
+
+By the contract, users should indicate the end of _parallel block_ for each
+previous start of _parallel block_.<br>
+Let's introduce RAII scoped object that will help to manage the contract.
+
+If the end of the parallel block is not indicated by the user, it will be done automatically when
+the last public reference is removed from the arena (i.e., task_arena is destroyed or a thread
+is joined for an implicit arena). This ensures correctness is
+preserved (threads will not be retained forever).
+
+## Considerations
+
+The alternative approaches were also considered.<br>
+We can express this state machine as complete graph and provide low-level interface that
+will give control over state transition.
+
+<img src="alternative_proposal.png" width=600>
+
+We considered this approach too low-level. Plus, it leaves a question: "How to manage concurrent changes of the state?".
+
+The retaining of worker threads should be implemented with care because
+it might introduce performance problems if:
+* Threads cannot migrate to another arena because they are
+  retained in the current arena.
+* Compute resources are not homogeneous, e.g., the CPU is hybrid.
+  Heavier involvement of less performant core types might result in artificial work
+  imbalance in the arena.
+
+
+## Open Questions in Design
+
+Some open questions that remain:
+* Are the suggested APIs sufficient?
+* Are there additional use cases that should be considered that we missed in our analysis?
+* Do we see any value if arena potentially can transition from one to another state?
+  * What if different types of workloads are mixed in one application?
+  * What if there concurrent calls to this API?
```

Comment:
See my comment above about making the approach to be a bit more generic. Essentially, I think we can write something like "implementation-defined" in case of a concurrent calls to this API. However, it seems to me that the behavior should be kind of relaxed, so to say. Meaning that if there is at least one "delayed leave" request happening concurrently with possibly a number of "fast leave" requests, then it, i.e., "delayed leave" policy prevails.

Also, having the request stated up front allows scheduler to know the runtime situation earlier, hence making better decisions about optimality of the workers' behavior.

Reply:
I think the last-block-to-end is a good way to break any conflicts, and this is the currently proposed behavior. If the last to end is "with_fast_leave", there's is some conflicting runtime use about to happen, so seems like the right thing to do is fast leave. If the last to end is not "with_fast_leave" on an arena that was created with "automatic" that means "automatic", so it still could opt to leave fast. As I read the state diagram, it doesn't really say it has to be a delayed leave if the final decrement was from a non fast leave end. So this looks like something to evaluate during implementation phase.